llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -7985,6 +7985,12 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
     return;
   }
   case Intrinsic::experimental_vector_partial_reduce_add: {
+
+    if (!TLI.shouldExpandPartialReductionIntrinsic(&I)) {
```

Comment:
What was the reasoning for expanding the intrinsic in SelectionDAGBuilder, rather than mapping to a new ISD node that would get expanded to some default implementation when targets don't do custom lowering?

That way, you wouldn't need to write such a complex `shouldExpandPartialReductionIntrinsic` function.

Reply:
I'm not sure what the reasoning was but I will look into creating an SDNode for this purpose. I do think it would be cleaner to have a default expansion and allow targets to lower it themselves.